### ISSUE TITLE
Some project names cause generated apps to not work

### ISSUE THREAD

**Issue by mrlacey** [Microsoft/WindowsTemplateStudio#1784]
Some project names cause problems when using the framework of the same name.

- Called **Prism** and uses the `Prism` framework causes the generated app to crash when started.
- Called **MVVMLight** and uses the `MVVMLight` framework causes NuGet restore to fail.
- Called **Caliburn.Micro** and uses the `Caliburn.Micro` framework causes NuGet restore to fail. (**Caliburn** as a name is fine.)

I forget and use these sometimes when making multiple similar versions of an app for debug/testing. Others may use them (for testing or otherwise) and it would be good to avoid them having issues.

Including these names in the `ReservedNamesValidator` should avoid the issue.

There are potentially lots of other project names that could be used that would cause errors but I don't think it's worth going looking for them. We can add any more when encountered but there are better ways to spend time right now.

**Comment by crutkas** [comment]
Good bug!

**Comment by ralarcon** [comment]
Agree, @mrlacey!

**Comment by mrlacey** [comment]
I thought this was going to be a simple fix, but it isn't.

The wizard only does validation on item names, not on the project name (which becomes the default namespace.) The project name is selected before the wizard starts.

I see four options available

Option 1. Modify the project name so it doesn't have issues.     
This could be done by appending a value to the name (e.g. "Prism" would become "Prism1" or "Prism_", etc.) The downside here is we'd be changing the value they want but without explaining why and they may not notice.

Option 2. Try and workaround the issue without changing the project name.   
This could be done by altering the code generation to append something to the name where it's used as part of a namespace.  
The downsides to this are that: we may end up doing something non-obvious with the generated code; it would make right-click>Add support more complicated (though needing to determine what, if anything should be done to make the root namespace different from the project name); it would also, I think, require some modifications to the current templates as some use the `Param_RootNamespace` and `Param_ItemNamespace` interchangeably but this would not always be possible in this scenario. This would also add further potential complication to template authoring.

Option 3. Prevent the generation of a project that has a potentially problematic name.  
Display an error message and prevent the wizard from creating an app.   
An error message something like "The project name you have chosen is likely to cause namespace collisions or other issues with creating a working project. Please chose a different name and try again."

Option 4. Like option 3 but just give a warning, and any issues are up to them to resolve.  
This feels like a cop-out. I also think it is likely to create more support issues where something gets generated, doesn't work correctly, and then we spend ages trying to track down the problem only for them to claim they didn't see the message.



My preference is for option 3.  This keeps things simple, prevents further issues in future, discourages people from doing something that is bad, and only requires minimal changes to the wizard.  
I can't see a good reason that someone _MUST_ use the name of a referenced assembly as the name of the project/app/exe. After all, if you were to call your project "System" you'd be right to expect some possibility for issues when it referenced the System dll.

(As this is pretty low priority-it's only a theoretical issue at the moment as no-one other than me has reported it-and it modifies the wizard, it could be done after the wizard redesign work.)

Thoughts? @crutkas? @ralarcon?

**Comment by crutkas** [comment]
i feel like #3 is in line when someone does this already, no?

**Comment by mrlacey** [comment]
> i feel like #3 is in line when someone does a non-valid name already, no?

No.

The current behavior is the app is created but you don't know there's a problem until you try and build it or run it.  
The build error looks like `Cycle detected. MVVMLight -> MvvmLight (>= 5.3.0).`. Which makes it look like a NuGet reference error at best.    
The runtime error is a win32 crash the debugger can't handle and so the app closes with no useful info.

What I'm suggesting with option 3 is that when the wizard is launched it will prevent the creation of a project if the name is potentially problematic. This way we will avoid generating a project that we know will have lots of issues.

**Comment by ralarcon** [comment]
You are right @mrlacey, the simplest way to catch this kind of errors is to block the "continuation" when a potentially unsafe name for the project is used. 

Regarding the urgency, I think it is not urgent, so we can introduce it once the new design is in place. Somehow, is like if you select a potentially dangerous name for your own project (i.e. use a project name called System for a class library). The only difference is that as we generate code automatically that probably you'll never manually create as you'll find errors early.

**Comment by mrlacey** [comment]
Moving to backlog

**Comment by mrlacey** [comment]
X-Ref #2892 
This should also catch and warn about project names that contain '$' (and possibly also '^')

Bumping this from #2892 as it requires a potential UI change in the wizard and I don't know how best to do that.

**Comment by mvegaca** [comment]
Fixed on #2977 
 - Naming markdown doc updated.
 - The error is cached before launching the wizard and is reported using a message box, that comes back to the new project window.

![image](https://user-images.githubusercontent.com/5556784/53802754-ee4e2180-3f42-11e9-8b39-1d108d0ad88f.png)

**Comment by sibille** [comment]
Verified in dev-nightly:
Templates version: 0.20.19073.1
Wizard version: 0.20.19073.1
